swarm/storage/dbstore.go

```diff
@@ -320,44 +325,66 @@ func (s *DbStore) Put(chunk *Chunk) {
 	var index dpaDBIndex
 
 	if s.tryAccessIdx(ikey, &index) {
-		if chunk.dbStored != nil {
-			close(chunk.dbStored)
-		}
-		log.Trace(fmt.Sprintf("Storing to DB: chunk already exists, only update access"))
+		log.Trace(fmt.Sprintf("DbStore: chunk already exists, only update access"))
 		return // already exists, only update access
 	}
+	s.doPut(chunk, ikey, &index)
+}
 
+// force putting into db, does not check access index
+func (s *DbStore) doPut(chunk *Chunk, ikey []byte, index *dpaDBIndex) {
 	data := encodeData(chunk)
-	//data := ethutil.Encode([]interface{}{entry})
-
-	if s.entryCnt >= s.capacity {
-		s.collectGarbage(gcArrayFreeRatio)
-	}
-
-	batch := new(leveldb.Batch)
-
-	batch.Put(getDataKey(s.dataIdx), data)
+	s.batch.Put(getDataKey(s.dataIdx), data)
 
 	index.Idx = s.dataIdx
-	s.updateIndexAccess(&index)
+	s.updateIndexAccess(index)
 
-	idata := encodeIndex(&index)
-	batch.Put(ikey, idata)
+	idata := encodeIndex(index)
+	s.batch.Put(ikey, idata)
 
-	batch.Put(keyEntryCnt, U64ToBytes(s.entryCnt))
 	s.entryCnt++
-	batch.Put(keyDataIdx, U64ToBytes(s.dataIdx))
 	s.dataIdx++
-	batch.Put(keyAccessCnt, U64ToBytes(s.accessCnt))
 	s.accessCnt++
-
-	s.db.Write(batch)
-	if chunk.dbStored != nil {
-		close(chunk.dbStored)
+	chunk.dbStored = s.batchC
+	select {
+	case s.batchesC <- struct{}{}:
+	default:
 	}
 	log.Trace(fmt.Sprintf("DbStore.Put: %v. db storage counter: %v ", chunk.Key.Log(), s.dataIdx))
 }
 
+func (s *DbStore) writeBatches() {
+	for range s.batchesC {
+		s.lock.RLock()
+		b := s.batch
+		e := s.entryCnt
+		d := s.dataIdx
+		a := s.accessCnt
+		c := s.batchC
+		s.batchC = make(chan bool)
+		s.batch = new(leveldb.Batch)
+		s.lock.RUnlock()
```

Comment:
This code is assigning to fields on `DbStore` so shouldn't it hold the write lock?